File: pbark/pbark/parser/parse_expression.py
from __future__ import annotations

from dataclasses import dataclass
from enum import Enum, auto

from pbark.parser.expression.comparison_op import ComparisonOp


class BinaryOp(Enum):
    PLUS = auto()
    MINUS = auto()
    STAR = auto()
    SLASH = auto()


class LogicalOp(Enum):
    AND = auto()
    OR = auto()
    NOR = auto()


class StashPart(Enum):
    ALL = auto()
    COUNT = auto()
    ELEMENT = auto()


class PilePart(Enum):
    ALL = auto()
    COUNT = auto()
    TOP = auto()


class ValueType(Enum):
    NUMBER = auto()
    WORDS = auto()
    NOTHING = auto()


@dataclass(frozen=True)
class StringLiteral:
    value: str


@dataclass(frozen=True)
class NumberLiteral:
    value: float


@dataclass(frozen=True)
class BooleanLiteral:
    value: bool


@dataclass(frozen=True)
class PeriodLiteral:
    pass


@dataclass(frozen=True)
class Empty:
    pass


@dataclass(frozen=True)
class NullLiteral:
    pass


@dataclass(frozen=True)
class Variable:
    value: str


@dataclass(frozen=True)
class Field:
    subject: str
    topic: str


@dataclass(frozen=True)
class Binary:
    left: ParseExpression
    op: BinaryOp
    right: ParseExpression


@dataclass(frozen=True)
class Not:
    operand: ParseExpression


@dataclass(frozen=True)
class Comparison:
    left: ParseExpression
    op: ComparisonOp
    right: ParseExpression


@dataclass(frozen=True)
class HasExact:
    subject: str
    topic: str
    amount: float


@dataclass(frozen=True)
class HasTrait:
    subject: str
    trait: str


@dataclass(frozen=True)
class Logical:
    left: ParseExpression
    op: LogicalOp
    right: ParseExpression


@dataclass(frozen=True)
class StashAccess:
    stash: str
    part: StashPart
    which: str | None = None


@dataclass(frozen=True)
class PileAccess:
    pile: str
    part: PilePart


@dataclass(frozen=True)
class Join:
    stash: str
    delimiter: ParseExpression


@dataclass(frozen=True)
class FunctionCall:
    name: str
    args: list[ParseExpression]


@dataclass(frozen=True)
class Length:
    value: ParseExpression


@dataclass(frozen=True)
class Contains:
    haystack: ParseExpression
    needle: ParseExpression


@dataclass(frozen=True)
class TypeCheck:
    value: ParseExpression
    type: ValueType


ParseExpression = (
    StringLiteral
    | NumberLiteral
    | BooleanLiteral
    | PeriodLiteral
    | Empty
    | NullLiteral
    | Variable
    | Field
    | Binary
    | Not
    | Comparison
    | HasExact
    | HasTrait
    | Logical
    | StashAccess
    | PileAccess
    | Join
    | FunctionCall
    | Length
    | Contains
    | TypeCheck
)

File: src/main/java/dev/klomptech/jbark/parser/ParseExpression.java
package dev.klomptech.jbark.parser;

import dev.klomptech.jbark.parser.expression.ComparisonOp;
import java.util.List;

public sealed interface ParseExpression
    permits ParseExpression.StringLiteral,
        ParseExpression.NumberLiteral,
        ParseExpression.BooleanLiteral,
        ParseExpression.PeriodLiteral,
        ParseExpression.Empty,
        ParseExpression.NullLiteral,
        ParseExpression.Variable,
        ParseExpression.Field,
        ParseExpression.Binary,
        ParseExpression.Not,
        ParseExpression.Comparison,
        ParseExpression.HasExact,
        ParseExpression.HasTrait,
        ParseExpression.Logical,
        ParseExpression.StashAccess,
        ParseExpression.PileAccess,
        ParseExpression.Join,
        ParseExpression.FunctionCall,
        ParseExpression.Length,
        ParseExpression.Contains,
        ParseExpression.TypeCheck {

  enum BinaryOp {
    PLUS,
    MINUS,
    STAR,
    SLASH
  }

  enum LogicalOp {
    AND,
    OR,
    NOR
  }

  enum StashPart {
    ALL,
    COUNT,
    ELEMENT
  }

  enum PilePart {
    ALL,
    COUNT,
    TOP
  }

  enum ValueType {
    NUMBER,
    WORDS,
    NOTHING
  }

  record StringLiteral(String value) implements ParseExpression {}

  record NumberLiteral(double value) implements ParseExpression {}

  record BooleanLiteral(boolean value) implements ParseExpression {}

  record PeriodLiteral() implements ParseExpression {}

  record Empty() implements ParseExpression {}

  record NullLiteral() implements ParseExpression {}

  record Variable(String value) implements ParseExpression {}

  record Field(String subject, String topic) implements ParseExpression {}

  record Binary(ParseExpression left, BinaryOp op, ParseExpression right)
      implements ParseExpression {}

  record Not(ParseExpression operand) implements ParseExpression {}

  record Comparison(ParseExpression left, ComparisonOp op, ParseExpression right)
      implements ParseExpression {}

  record HasExact(String subject, String topic, double amount) implements ParseExpression {}

  record HasTrait(String subject, String trait) implements ParseExpression {}

  record Logical(ParseExpression left, LogicalOp op, ParseExpression right)
      implements ParseExpression {}

  record StashAccess(String stash, StashPart part, String which) implements ParseExpression {}

  record PileAccess(String pile, PilePart part) implements ParseExpression {}

  record Join(String stash, ParseExpression delimiter) implements ParseExpression {}

  record FunctionCall(String name, List<ParseExpression> args) implements ParseExpression {}

  record Length(ParseExpression value) implements ParseExpression {}

  record Contains(ParseExpression haystack, ParseExpression needle) implements ParseExpression {}

  record TypeCheck(ParseExpression value, ValueType type) implements ParseExpression {}
}
